### PR TITLE
Full Inbound Logic + Lambdas

### PR DESCRIPTION
**Pull request by jonahkaye**
Ticket: #[2223](https://github.com/metriport/metriport-internal/issues/1766)

### Dependencies

- downstream: https://github.com/metriport/metriport-internal/pull/1964
- downstream: https://github.com/metriport/metriport/pull/2406

### Description

- inbound schemas, xml processing, and xml response generation 
-  new lambdas with full inbound logic 

### Testing

- Local
  - [x] lots of unit testing
  

### Release Plan

- Update urls in IHE directory to point to v2 endpoints
- Ship downstream PR to remove all ihev1 infra 
- [ ] Merge this

### CODE REVIEW

File: packages/ihe-gateway-sdk/src/models/shared.ts
Comment (jonahkaye) [inline packages/ihe-gateway-sdk/src/models/shared.ts:1]: adding the signature to both request and response so we can return it

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/xcpd/process/xcpd-request.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    city: addr.value.city ? String(addr.value.city) : undefined,
+    state: addr.value.state ? String(addr.value.state) : undefined,
+    postalCode: addr.value.postalCode ? String(addr.value.postalCode) : undefined,
+    country: addr.value.country ? String(addr.value.country) : undefined,
+  }));
+
+  // is it a phone or an email
+  const telecom = toArray(queryParams.patientTelecom).map(tel => ({
+    system: "phone",
+    value: tel.value._value,
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/inbound/xcpd/process/xcpd-request.ts:33]: maybe @thomasyopes knows if we have a function to determine this already?

File: packages/core/src/external/carequality/error.ts [hunk 2]
```diff
@@ -78,6 +78,7 @@ function constructBaseErrorResponse(
     patientId: payload.patientId,
     timestamp: payload.timestamp,
     responseTimestamp: new Date().toISOString(),
+    signatureConfirmation: payload.signatureConfirmation,
```
Comment (leite08) [inline packages/core/src/external/carequality/error.ts:81]: Is this fine to send to Sentry and/or log (re: sensitive information)?

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/__tests__/process-dq-request.test.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    });
+    if (iti38Response.documentReference && iti38Response.documentReference[0]) {
+      expect(extractDocumentUniqueId(iti38Response.documentReference[0].docUniqueId)).toEqual(
+        docUniqueId
+      );
+    } else {
+      throw new Error(
+        "iti38Response.documentReference is undefined or has wrong document unique id"
+      );
+    }
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/inbound/__tests__/process-dq-request.test.ts:79]: Not a huge deal, but I'd keep the main flow/indentation as part of the test, and have alternative flows on conditionals, like

```
... // prepare/execute
if (unexpected condition) error;
... // check results

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/shared.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+export function extractTimestamp(header: SamlHeader): string {
+  return header.Security.Timestamp.Created;
+}
+
+export function createSecurityHeader({
+  signatureConfirmation,
+}: {
+  signatureConfirmation?: string | undefined;
+}) {
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/inbound/shared.ts:112]: NABD if not, but don't we have a type to represent the return of this function?

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/create/dq-response.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    "soap:Envelope": {
+      "@_xmlns:soap": namespaces.soap,
+      "@_xmlns:wsa": namespaces.wsa,
+      "soap:Header": {
+        ...securityHeader,
+        "wsa:Action": {
+          "#text": "urn:ihe:iti:2007:CrossGatewayQueryResponse",
+          "@_mustUnderstand": "1",
+        },
+        "wsa:MessageID": wrapIdInUrnUuid(uuidv4()), // TODO track this
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/create/dq-response.ts:74]: nit: is the TODO part of a follow-up/master ticket? Useful to add the ticket number after `TODO`

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/process/dq-request.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      id: stripUrnPrefix(extractText(iti38Request.Envelope.Header.MessageID)),
+      timestamp: extractTimestamp(iti38Request.Envelope.Header),
+      samlAttributes,
+      externalGatewayPatient,
+      signatureConfirmation: extractText(
+        iti38Request.Envelope.Header.Security.Signature.SignatureValue
+      ),
+    };
+  } catch (error) {
+    console.log(error);
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/process/dq-request.ts:54]: left-over?

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/process/dr-request.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  try {
+    const parser = new XMLParser({
+      ignoreAttributes: false,
+      attributeNamePrefix: "_",
+      textNodeName: "_text",
+      parseAttributeValue: false,
+      removeNSPrefix: true,
+    });
+    const jsonObj = parser.parse(request);
+    console.log("jsonObj", JSON.stringify(jsonObj, null, 2));
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/process/dr-request.ts:27]: left-over?

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/process/dr-request.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+      id: stripUrnPrefix(extractText(iti39Request.Envelope.Header.MessageID)),
+      timestamp: extractTimestamp(iti39Request.Envelope.Header),
+      samlAttributes,
+      documentReference,
+      signatureConfirmation: extractText(
+        iti39Request.Envelope.Header.Security.Signature.SignatureValue
+      ),
+    };
+  } catch (error) {
+    console.log(error);
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/process/dr-request.ts:45]: ditto: left-over? here and other `console.log` calls

generally speaking:
- to log, let's use `out.log`
- when logging, no new-lines
- log if needed (so logging on the handling code is useful if we're not logging upstream - which I assume we are and also capturing it)

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/xcpd/create/xcpd-response.ts [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+): object {
+  const { ack, queryResponseCode } = createAckAndQueryResponseCode(response);
+  const queryByParameter = createQueryByParameter(request);
+  const subject = response.patientMatch ? createSubjectAndRegistrationEvent(response) : undefined;
+
+  const soapBody = {
+    "@_xmlns:urn": namespaces.hl7,
+    PRPA_IN201306UV02: {
+      id: {
+        "@_root": uuidv7(), // TODO change this if we ever end up keeping track of these.
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/inbound/xcpd/create/xcpd-response.ts:188]: ditto: how to keep track of this item after this PR is merged.

nit: why the diff approach to UUID generation, v4 vs v7 in diff places of the PR?

File: packages/core/src/external/carequality/ihe-gateway-v2/schema.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+  z.object({
+    PurposeOfUse: codeSchema,
+  }),
+  textSchema,
+]);
+export type AttributeValue = z.infer<typeof AttributeSchema>;
+
+export const samlHeaderSchema = z.object({
+  // messageId
+  // signature
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/schema.ts:82]: left-over?

File: packages/infra/lib/ihe-stack.ts [hunk 11]
```diff
@@ -183,6 +187,43 @@ export class IHEStack extends Stack {
...
+
+    const patientDiscoveryLambdaV2 = this.setupPatientDiscoveryLambda({
+      props,
+      lambdaLayers,
+      vpc,
+      secrets,
+      posthogSecretName,
+      alarmSnsAction,
+      name: "IHEInboundPatientDiscoveryV2",
+      entry: "ihe-gateway-v2-inbound-patient-discovery",
```
Comment (leite08) [inline packages/infra/lib/ihe-stack.ts:202]: nit: I imagine these are optional until we remove the old code; could make the values for the new ones the default of the parameter so we don't need to change this code afterwards. But defs NABD if we revert this later to not pass these as parameters once we don't have the v1 anymore.

File: packages/lambdas/src/ihe-gateway-v2-inbound-document-query.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+import { createInboundDqResponse } from "@metriport/core/external/carequality/ihe-gateway-v2/inbound/xca/create/dq-response";
+import { analyticsAsync, EventTypes } from "@metriport/core/external/analytics/posthog";
+import { getEnvVarOrFail, getEnvVar } from "@metriport/core/util/env-var";
+
+const apiUrl = getEnvVarOrFail("API_URL");
+const region = getEnvVarOrFail("AWS_REGION");
+const engineeringCxId = getEnvVar("ENGINEERING_CX_ID");
+const postHogSecretName = getEnvVar("POST_HOG_API_KEY_SECRET");
+
+export const handler = Sentry.AWSLambda.wrapHandler(async (event: ALBEvent) => {
```
Comment (leite08) [inline packages/lambdas/src/ihe-gateway-v2-inbound-document-query.ts:16]: I suggest we create lambdas from now on with the regular, non-Sentry handler, and always add a global try/catch. The reasoning is that our own global try/catch can send more context info to Sentry than the default one.

See `sqs-to-converter.ts` for reference.

Although, maybe we shouldn't send these to Sentry at all - inbound is our of our control and could kill our quotas. Could have a CW alarm based on the ratio of lambda invocation errors.

Comment (leite08) [reply packages/lambdas/src/ihe-gateway-v2-inbound-document-query.ts:16]: Odd to have the event type as `ALBEvent` - I'd think this type would be used for requests sent from the ALB to the lambda.

File: packages/core/src/external/carequality/error.ts (hunk 2 above)
Comment (jonahkaye) [reply packages/core/src/external/carequality/error.ts:81]: Its not sensitive

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/shared.ts (hunk 4 above)
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/inbound/shared.ts:112]: we dont. I can add the object return type

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/process/dr-request.ts (hunk 8 above)
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/process/dr-request.ts:45]: sorry about these. I just added them for when I was testing on the lambda just now

File: packages/infra/lib/ihe-stack.ts (hunk 11 above)
Comment (jonahkaye) [reply packages/infra/lib/ihe-stack.ts:202]: I will change this in the downstream PR to not pass as parameters

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/xcpd/create/xcpd-response.ts (hunk 9 above)
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/inbound/xcpd/create/xcpd-response.ts:188]: changed to v4

File: packages/lambdas/src/ihe-gateway-v2-inbound-document-query.ts (hunk 12 above)
Comment (jonahkaye) [reply packages/lambdas/src/ihe-gateway-v2-inbound-document-query.ts:16]: Decided to not send to sentry at all. Returning 500s for now. 

ALBEvent is the type for for ALB connecting to Lambda, which is what we have

File: packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/process/dr-request.ts (hunk 8 above)
Comment (leite08) [reply packages/core/src/external/carequality/ihe-gateway-v2/inbound/xca/process/dr-request.ts]: No worries! I noticed there are some still left over. Please search for `console.` on this PR and you'll see:
- [a log that seems for debug on the lambda](https://github.com/metriport/metriport/pull/2223/files#diff-508814c9bba3f797030b850d8e9fc603e6dfbfde2a96fb0f581880cc5189ca06R57)
- [a multi-line log on core](https://github.com/metriport/metriport/pull/2223/files#diff-9d604ef2ec647d4b0fc86313f850a472b6dd539d1cac759b05516dba0eac41e9R76)

File: packages/lambdas/src/ihe-gateway-v2-inbound-document-query.ts (hunk 12 above)
Comment (leite08) [reply packages/lambdas/src/ihe-gateway-v2-inbound-document-query.ts]: I thought the inbound lambda was being called straight from the API GW, no?

Comment (jonahkaye) [reply packages/lambdas/src/ihe-gateway-v2-inbound-document-query.ts]: Sounds good on the metrics. 
Your totally right on the AlbEvent thing. I confused myself here. I will find the proper type and replace it

File: packages/infra/lib/ihe-stack/ihe-gateway.ts [hunk 13]
```diff
@@ -106,4 +112,22 @@ export function createIHEGateway(stack: Construct, props: IHEGatewayProps): void
       ),
     }),
   });
+
+  apiGateway.addRoutes({
+    path: "/v2/patient-discovery",
+    methods: [apigwv2.HttpMethod.POST],
+    integration: new HttpLambdaIntegration("IHEGWPDIntegrationV2", props.patientDiscoveryLambdaV2),
+  });
```
Comment (jonahkaye) [inline packages/infra/lib/ihe-stack/ihe-gateway.ts:120]: new urls

**Pull request merged**